src/gateway/GatewayV1.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.13;
+
+import "./Gateway.sol";
+
+interface IBalancerFlashLoan {
+    function flashLoan(
+        address recipient,
+        IERC20[] memory tokens,
+        uint256[] memory amounts,
+        bytes memory userData
+    ) external;
+}
+
+/// @title ECG Gateway V1
+/// @notice Gateway to interract via multicall with the ECG
+/// Owner can select which calls are allowed
+/// @custom:feature flashloan from balancer vault
+contract GatewayV1 is Gateway {
+    /// @notice Address of the Balancer Vault, used for initiating flash loans.
+    address public immutable balancerVault =
+        0xBA12222222228d8Ba445958a75a0704d566BF2C8;
+
+    function multicallWithBalancerFlashLoan(
+        IERC20[] calldata tokens,
+        uint256[] calldata amounts,
+        bytes[] calldata calls // Calls to be made after receiving the flash loan
+    ) public whenNotPaused {
+        require(
+            _originalSender == address(1),
+            "Gateway: original sender already set"
+        );
+
+        _originalSender = msg.sender;
+        // Clear existing StoredCalls
+        delete _storedCalls;
+
+        // Manually copy each element
+        for (uint i = 0; i < calls.length; i++) {
+            _storedCalls.push(calls[i]);
+        }
+
+        // Initiate the flash loan
+        IBalancerFlashLoan(balancerVault).flashLoan(
+            address(this),
+            tokens,
+            amounts,
+            ""
```

Comment:
why do we store calls in storage instead of passing these here & reading from userData in receiveFlashloan?

Reply:
Hmmmm I think we did it to be sure that it was the original msg.sender that had stored the calls.

Might not be needed now as we test original sender.

I think, because the balancer vault address is hardcoded, that it's not risky to just give the args to balancer for it to git it back in the receiveFlashloan